src/main/java/com/cleanroommc/groovyscript/compat/mods/prodigytech/ExplosionFurnaceAdditives.java

```diff
@@ -0,0 +1,115 @@
+package com.cleanroommc.groovyscript.compat.mods.prodigytech;
+
+import com.cleanroommc.groovyscript.api.IIngredient;
+import com.cleanroommc.groovyscript.api.documentation.annotations.Example;
+import com.cleanroommc.groovyscript.api.documentation.annotations.MethodDescription;
+import com.cleanroommc.groovyscript.api.documentation.annotations.RegistryDescription;
+import com.cleanroommc.groovyscript.helper.ingredient.OreDictIngredient;
+import com.cleanroommc.groovyscript.registry.VirtualizedRegistry;
+import lykrast.prodigytech.common.recipe.ExplosionFurnaceManager;
+import net.minecraft.item.ItemStack;
+
+class EFAdditiveRecipe {
+    boolean isExplosive;
+    IIngredient input;
+    int value;
+
+    EFAdditiveRecipe(boolean isExplosive, IIngredient input, int value) {
+        this.isExplosive = isExplosive;
+        this.input = input;
+        this.value = value;
+    }
+
+    void register() {
+        if (this.isExplosive) {
+            if (this.input instanceof OreDictIngredient) {
+                ExplosionFurnaceManager.addExplosive(((OreDictIngredient) this.input).getOreDict(), this.value);
+            } else {
+                for (ItemStack it : this.input.getMatchingStacks()) {
+                    ExplosionFurnaceManager.addExplosive(it, this.value);
+                }
+            }
+        } else {
+            if (this.input instanceof OreDictIngredient) {
+                ExplosionFurnaceManager.addDampener(((OreDictIngredient) this.input).getOreDict(), this.value);
+            } else {
+                for (ItemStack it : this.input.getMatchingStacks()) {
+                    ExplosionFurnaceManager.addDampener(it, this.value);
+                }
+            }
+        }
+    }
+
+    void unregister() {
+        if (this.isExplosive) {
+            if (this.input instanceof OreDictIngredient) {
+                ExplosionFurnaceManager.removeExplosive(((OreDictIngredient) this.input).getOreDict());
+            } else {
+                for (ItemStack it : this.input.getMatchingStacks()) {
+                    ExplosionFurnaceManager.removeExplosive(it);
+                }
+            }
+        } else {
+            if (this.input instanceof OreDictIngredient) {
+                ExplosionFurnaceManager.removeDampener(((OreDictIngredient) this.input).getOreDict());
+            } else {
+                for (ItemStack it : this.input.getMatchingStacks()) {
+                    ExplosionFurnaceManager.removeDampener(it);
+                }
+            }
+        }
+    }
+}
+
+@RegistryDescription
+public class ExplosionFurnaceAdditives extends VirtualizedRegistry<EFAdditiveRecipe> {

```

Comment:
can we get a `removeAll` for both explosives and dampeners?

Reply:
So here, I have a very bizarre issue I am not sure if it is fixable,

so you see, the Explosive's ItemStack is only exposed as a private final property. As far as I know this means it is getting optimized by the JVM in the way where even mixins can't access it. This means `removeAllExplosives()` and such won't be possible to roll back.